odl/util/normalize.py

```diff
@@ -0,0 +1,193 @@
+
+import numpy as np
+
+
+__all__ = ('normalized_index_expression', 'normalized_scalar_param_list')
+
+
+def normalized_scalar_param_list(param, length, param_conv=None,
+                                 keep_none=True):
+    """Return a list of given length from a scalar parameter.
+
+    The typical use case is when a single value or a sequence of
+    values is accepted as input. This function makes a list from
+    a given sequence or a list of identical elements from a single
+    value, with cast to a given parameter type if desired.
+
+    To distinguish a parameter sequence from a single parameter, the
+    following rules are applied:
+
+    * If ``param`` is not a sequence, it is treated as a single
+    parameter (e.g. ``1``).
+    * If ``len(param) == length == 1``, then ``param`` is interpreted
+      as a single parameter (e.g. ``[1]`` or ``'1'``).
+    * If ``len(param) == length != 1``, then ``param`` is interpreted as
+      sequence of parameters.
+    * Otherwise, ``param`` is interpreted as a single parameter.
+
+    Note that this function is not applicable to parameters which are
+    themselves iterable (e.g. ``'abc'`` with ``length=3`` will be
+    interpreted as equivalent to ``['a', 'b', 'c']``).
+
+    Parameters
+    ----------
+    param :
+        Input parameter to turn into a list.
+    length : positive int
+        Desired length of the output list.
+    param_conv : callable, optional
+        Conversion applied to each list element. None means no conversion.
+    keep_none : bool, optional
+        If True, None is not converted.
+
+    Returns
+    -------
+    plist : list
+        Input parameter turned into a list of length ``length``.
+    """
+    length, length_in = int(length), length
+    if length <= 0:
+        raise ValueError('`length` must be positive, got {}'.format(length_in))
+
+    try:
+        # TODO: always use this when numpy >= 1.10 can be assumed
+        out_list = list(np.broadcast_to(param, (length,)))
+    except AttributeError:
+        try:
+            param_len = len(param)
+        except TypeError:
+            # Not a sequence -> single parameter
+            out_list = [param] * length
+        else:
+            if param_len == 1:
+                out_list = list(param) * length
+            elif param_len == length:
+                out_list = list(param)
+            else:
+                raise ValueError('sequence `param` has length {}, expected {}'
+                                 ''.format(param_len, length))
+
+    if len(out_list) != length:
+        raise ValueError('sequence `param` has length {}, expected {}'
+                         ''.format(len(out_list), length))
+
+    if param_conv is not None:
+        try:
+            for i, p in enumerate(out_list):
+                if p is None and keep_none:
+                    pass
+                else:
+                    out_list[i] = param_conv(p)
+        except TypeError as error:
```

Comment:
I would prefer simply letting this error trickle through. In python 2 raise_from kills the call stack making error searching hard.